File: TestNeuralNetwork.java
import org.junit.jupiter.api.Assertions;
import org.ejml.simple.SimpleMatrix;
import org.junit.jupiter.api.Test;
import org.ejml.EjmlUnitTests;

class TestNeuralNetwork {

    @Test
    void test_sigmoid() {
        // Create arrays for input matrices
        double[][] test_arr_inp_0 = new double[][]{new double[]{-5d}};
        double[][] test_arr_inp_1 = new double[][]{new double[]{0d}};
        double[][] test_arr_inp_2 = new double[][]{new double[]{5d}};
        double[][] test_arr_inp_3 = new double[][]{new double[]{4d, 5d, 6d}};
        double[][] test_arr_inp_4 = new double[][]{
                                        new double[]{-1d},
                                        new double[]{0d},
                                        new double[]{1d}
        };
        double[][] test_arr_inp_5 = new double[4][5];
        double init_val = -1.0;
        for(int col = 0; col < test_arr_inp_5[0].length; col++){
            for(int row = 0; row < test_arr_inp_5.length; row++){
                test_arr_inp_5[row][col] = init_val;
                init_val += 0.1;
            }
        }

        // Create arrays for output matrices
        double[][] test_arr_out_0 = new double[][]{ new double[]{0.006692851}};
        double[][] test_arr_out_1 = new double[][]{ new double[]{0.500000000}};
        double[][] test_arr_out_2 = new double[][]{ new double[]{0.993307149}};
        double[][] test_arr_out_3 = new double[][]{ new double[]{0.9820137900, 0.9933071490, 0.997527377}};
        double[][] test_arr_out_4 = new double[][]{ new double[]{0.268941421},
                                                    new double[]{0.500000000},
                                                    new double[]{0.731058579}
        };
        double[][] test_arr_out_5 = new double[][]{
                    new double[]{0.268941421, 0.354343694, 0.450166003, 0.549833997, 0.645656306},
                    new double[]{0.289050497, 0.377540669, 0.475020813, 0.574442517, 0.668187772},
                    new double[]{0.310025519, 0.401312340, 0.500000000, 0.598687660, 0.689974481},
                    new double[]{0.331812228, 0.425557483, 0.524979187, 0.622459331, 0.710949503}
        };

        // Create input SimpleMatrices
        SimpleMatrix test_inp_0 = new SimpleMatrix(test_arr_inp_0);
        SimpleMatrix test_inp_1 = new SimpleMatrix(test_arr_inp_1);
        SimpleMatrix test_inp_2 = new SimpleMatrix(test_arr_inp_2);
        SimpleMatrix test_inp_3 = new SimpleMatrix(test_arr_inp_3);
        SimpleMatrix test_inp_4 = new SimpleMatrix(test_arr_inp_4);
        SimpleMatrix test_inp_5 = new SimpleMatrix(test_arr_inp_5);

        // Create output SimpleMatrices
        SimpleMatrix test_out_0 = new SimpleMatrix(test_arr_out_0);
        SimpleMatrix test_out_1 = new SimpleMatrix(test_arr_out_1);
        SimpleMatrix test_out_2 = new SimpleMatrix(test_arr_out_2);
        SimpleMatrix test_out_3 = new SimpleMatrix(test_arr_out_3);
        SimpleMatrix test_out_4 = new SimpleMatrix(test_arr_out_4);
        SimpleMatrix test_out_5 = new SimpleMatrix(test_arr_out_5);

        // Create NeuralNetwork object
        NeuralNetwork nn = new NeuralNetwork();

        // Perform assertions
        double tolerance = 0.000000001;
        EjmlUnitTests.assertEquals(test_out_0.getDDRM(), nn.sigmoid(test_inp_0).getDDRM(), tolerance);
        EjmlUnitTests.assertEquals(test_out_1.getDDRM(), nn.sigmoid(test_inp_1).getDDRM(), tolerance);
        EjmlUnitTests.assertEquals(test_out_2.getDDRM(), nn.sigmoid(test_inp_2).getDDRM(), tolerance);
        EjmlUnitTests.assertEquals(test_out_3.getDDRM(), nn.sigmoid(test_inp_3).getDDRM(), tolerance);
        EjmlUnitTests.assertEquals(test_out_4.getDDRM(), nn.sigmoid(test_inp_4).getDDRM(), tolerance);
        EjmlUnitTests.assertEquals(test_out_5.getDDRM(), nn.sigmoid(test_inp_5).getDDRM(), tolerance);
    }

    @Test
    void test_nn_cost_function(){
        // Input Parameter Instantiation
        int num_inp_layers = 2;
        int num_hidden_layers = 2;
        int num_labels = 4;
        SimpleMatrix parameters  = new SimpleMatrix(new double[][]{
                new double[]{0.100, 0.200, 0.300, 0.400, 0.500, 0.600, 0.700,
                             0.800, 0.900, 1.000, 1.100, 1.200, 1.300, 1.400,
                             1.500, 1.600, 1.700, 1.800}
        });
        SimpleMatrix input_data = new SimpleMatrix(new double[][]{
                new double[]{0.5403023, -0.4161468},
                new double[]{-0.9899925, -0.6536436},
                new double[]{0.2836622, 0.9601703}
        });
        SimpleMatrix output_data = new SimpleMatrix(new double[][]{
                new double[]{4d},
                new double[]{2d},
                new double[]{3d}
        });
        int lambda = 4;

        //Output Data
        double unregularized_cost_output = 7.4069698;

        // Create NeuralNetwork object
        NeuralNetwork nn = new NeuralNetwork();

        // Unregularized Cost Function Test
        double tolerance = 0.0000001;
        Assertions.assertEquals(unregularized_cost_output, nn.nn_cost_function(
                parameters, input_data, output_data, num_inp_layers,
                num_hidden_layers, num_labels, 0 ), tolerance);

        return;
    }
}

File: RoguelikeCompanion/TestProject1/TestNeuralNetwork.cs
using System;
using Xunit;
using RoguelikeCompanion;
using MathNet.Numerics.LinearAlgebra;
using System.Drawing;

namespace TestRogulikeCompanion
{
    public class TestNeuralNetwork
    {
        [Fact]
        public void TestSigmoid()
        {
            // Create arrays for input matrices
            var testInp0 = Matrix<double>.Build.DenseOfArray( new double[,] { { -5D } });
            var testInp1 = Matrix<double>.Build.DenseOfArray( new double[,] { { 0D } });
            var testInp2 = Matrix<double>.Build.DenseOfArray( new double[,] { { 5D } });
            var testInp3 = Matrix<double>.Build.DenseOfArray( new double[,] { { 4D, 5D, 6D } });
            var testInp4 = Matrix<double>.Build.DenseOfArray( new double[,] { { -1D }, { 0D }, { 1D } });
            var testArr5 = new double[4,5];
            double val = -1.0;
            for (int col = 0; col < testArr5.GetLength(1); col++)
            {
                for (int row = 0; row < testArr5.GetLength(0); row++)
                {
                    testArr5[row, col] = val;
                    val += 0.1;
                }
            }
            var testInp5 = Matrix<double>.Build.DenseOfArray(testArr5);

            // Create arrays for output matrices
            var testOut0 = Matrix<double>.Build.DenseOfArray( new double[,] { { 0.006692851 } });
            var testOut1 = Matrix<double>.Build.DenseOfArray( new double[,] { { 0.500000000 } });
            var testOut2 = Matrix<double>.Build.DenseOfArray( new double[,] { { 0.993307149 } });
            var testOut3 = Matrix<double>.Build.DenseOfArray( new double[,] { { 0.9820137900, 0.9933071490, 0.997527377 } });
            var testOut4 = Matrix<double>.Build.DenseOfArray( new double[,] { {0.268941421}, {0.500000000}, {0.731058579}});
            var testOut5 = Matrix<double>.Build.DenseOfArray(new double[,]{
                                                        {0.268941421, 0.354343694, 0.450166003, 0.549833997, 0.645656306},
                                                        {0.289050497, 0.377540669, 0.475020813, 0.574442517, 0.668187772},
                                                        {0.310025519, 0.401312340, 0.500000000, 0.598687660, 0.689974481},
                                                        {0.331812228, 0.425557483, 0.524979187, 0.622459331, 0.710949503} });

            // Create testing tuple array
            Tuple<Matrix<double>, Matrix<double>>[] testingTuples =
            {
                Tuple.Create(testInp0, testOut0),
                Tuple.Create(testInp1, testOut1),
                Tuple.Create(testInp2, testOut2),
                Tuple.Create(testInp3, testOut3),
                Tuple.Create(testInp4, testOut4),
                Tuple.Create(testInp5, testOut5),
            };

            // Perform assertions, element wise with precision
            foreach(Tuple<Matrix<double>, Matrix<double>> newTestingPair in testingTuples)
            {
                double[] experimental = NeuralNetwork.sigmoid(newTestingPair.Item1).ToRowMajorArray();
                double[] actual = newTestingPair.Item2.ToRowMajorArray();
                for (int idx = 0; idx < experimental.Length; idx++)
                {
                    Assert.Equal(experimental[idx], actual[idx], 9);
                }
            }
            
        }


        [Fact]
        public void TestTransformCharArrayToString()
        {
            // Testing inputs
            string[] inp0 = { "" };
            string[] inp1 = { "T", "e", "s", "t" };
            string[] inp2 = { "S", "h", "a", "d", "e", "s", "ERRAP", "s", "SPACE", "r", "e", "v", "o", "l", "v", "e", "r" };
            string[] inp3 = { "D", "a", "r", "t", "SPACE", "g", "ERRU", "ERRN" };
            string[] inp4 = { "H", "u", "ERRN", "ERRT", "s", "m", "a", "n" };

            // Testing outputs
            string out0 = "";
            string out1 = "Test";
            string out2 = "Shades's_Revolver";
            string out3 = "Dart_G";
            string out4 = "Husman";

            // Assertions
            Assert.Equal(NeuralNetwork.transformCharArrayToString(inp0), out0);
            Assert.Equal(NeuralNetwork.transformCharArrayToString(inp1), out1);
            Assert.Equal(NeuralNetwork.transformCharArrayToString(inp2), out2);
            Assert.Equal(NeuralNetwork.transformCharArrayToString(inp3), out3);
            Assert.Equal(NeuralNetwork.transformCharArrayToString(inp4), out4);
        }

        [Fact]
        public static void TestNewImagePrediction()
        {
            // Testing inputs
            Bitmap inp0 = (Bitmap)Image.FromFile("C:/Users/Grant/Desktop/Java_Rouge_Like_App/screenshots/class_2_border/class_2_0.jpg");
            Bitmap inp1 = (Bitmap)Image.FromFile("C:/Users/Grant/Desktop/Java_Rouge_Like_App/screenshots/class_2_border/class_2_2.jpg");
            Bitmap inp2 = (Bitmap)Image.FromFile("C:/Users/Grant/Desktop/Java_Rouge_Like_App/screenshots/class_2_border/class_2_3.jpg");
            Bitmap inp3 = (Bitmap)Image.FromFile("C:/Users/Grant/Desktop/Java_Rouge_Like_App/screenshots/class_2_border/class_2_5.jpg");
            Bitmap inp4 = (Bitmap)Image.FromFile("C:/Users/Grant/Desktop/Java_Rouge_Like_App/screenshots/class_2_border/class_2_9.jpg");
            Bitmap inp5 = (Bitmap)Image.FromFile("C:/Users/Grant/Desktop/Java_Rouge_Like_App/screenshots/class_2_border/class_2_7.jpg");


            // Testing outputs
            string out0 = "Bullet_Time";
            string out1 = "Backpack";
            string out2 = "Ice_Cube";
            string out3 = "Orange";
            string out4 = "Trank_Gun";           
            string out5 = "Cloranthy_Ring";

            // Create Testing Tuples
            Tuple<Bitmap, string>[] testingTuple =
            {
                Tuple.Create(inp0, out0),
                Tuple.Create(inp1, out1),
                Tuple.Create(inp2, out2),
                Tuple.Create(inp3, out3),
                Tuple.Create(inp4, out4),
                Tuple.Create(inp5, out5)
            };

            // Perform testing
            NeuralNetwork nn = new NeuralNetwork();
            foreach(var pair in testingTuple)
            {
                Assert.Equal(nn.newImagePrediction(pair.Item1), pair.Item2);
            }
        }
    }
}
